include/mcl_3dl/state_6dof.h

```diff
@@ -0,0 +1,278 @@
+    RPYVec()
+    {
+    }
+    explicit RPYVec(const mcl_3dl::Vec3 &v)
+    {
+      this->v = v;
+    }
+    RPYVec(const float &r, const float &p, const float y)
+    {
+      this->v.x = r;
+      this->v.y = p;
+      this->v.z = y;
+    }
+  };
+  RPYVec rpy;
+  float &operator[](const size_t i)override
+  {
+    switch (i)
+    {
+      case 0:
+        return pos.x;
+      case 1:
+        return pos.y;
+      case 2:
+        return pos.z;
+      case 3:
+        return rot.x;
+      case 4:
+        return rot.y;
+      case 5:
+        return rot.z;
+      case 6:
+        return rot.w;
+      case 7:
+        return odom_err_integ_lin.x;
+      case 8:
+        return odom_err_integ_lin.y;
+      case 9:
+        return odom_err_integ_lin.z;
+      case 10:
+        return odom_err_integ_ang.x;
+      case 11:
+        return odom_err_integ_ang.y;
+      case 12:
+        return odom_err_integ_ang.z;
+    }
+    return pos.x;
+  }
+  const float &operator[](const size_t i) const
+  {
+    switch (i)
+    {
+      case 0:
+        return pos.x;
+      case 1:
+        return pos.y;
+      case 2:
+        return pos.z;
+      case 3:
+        return rot.x;
+      case 4:
+        return rot.y;
+      case 5:
+        return rot.z;
+      case 6:
+        return rot.w;
+      case 7:
+        return odom_err_integ_lin.x;
+      case 8:
+        return odom_err_integ_lin.y;
+      case 9:
+        return odom_err_integ_lin.z;
+      case 10:
+        return odom_err_integ_ang.x;
+      case 11:
+        return odom_err_integ_ang.y;
+      case 12:
+        return odom_err_integ_ang.z;
+    }
+    return pos.x;
+  }
+  size_t size() const override
+  {
+    return 10;
+  }
+  void normalize() override
+  {
+    rot.normalize();
+  }
+  State6DOF()
+  {
+    diff = false;
+    noise_ll_ = noise_la_ = noise_aa_ = noise_al_ = 0.0;
+    odom_err_integ_lin = mcl_3dl::Vec3(0.0, 0.0, 0.0);
+    odom_err_integ_ang = mcl_3dl::Vec3(0.0, 0.0, 0.0);
+  }
+  State6DOF(const mcl_3dl::Vec3 pos, const mcl_3dl::Quat rot)
+  {
+    this->pos = pos;
+    this->rot = rot;
+    noise_ll_ = noise_la_ = noise_aa_ = noise_al_ = 0.0;
+    odom_err_integ_lin = mcl_3dl::Vec3(0.0, 0.0, 0.0);
+    odom_err_integ_ang = mcl_3dl::Vec3(0.0, 0.0, 0.0);
+    diff = false;
+  }
+  State6DOF(const mcl_3dl::Vec3 pos, const mcl_3dl::Vec3 rpy)
+  {
+    this->pos = pos;
+    this->rpy = RPYVec(rpy);
+    noise_ll_ = noise_la_ = noise_aa_ = noise_al_ = 0.0;
+    odom_err_integ_lin = mcl_3dl::Vec3(0.0, 0.0, 0.0);
+    odom_err_integ_ang = mcl_3dl::Vec3(0.0, 0.0, 0.0);
+    diff = true;
+  }
+  bool isDiff()
+  {
+    return diff;
+  }
+  void transform(pcl::PointCloud<pcl::PointXYZI> &pc) const
+  {
+    auto r = rot.normalized();
+    for (auto &p : pc.points)
+    {
+      auto t = r * mcl_3dl::Vec3(p.x, p.y, p.z) + pos;
+      p.x = t.x;
+      p.y = t.y;
+      p.z = t.z;
+    }
+  }
+  static State6DOF generateNoise(
+      std::default_random_engine &engine_,
```

Comment:
Depending on compiler, default_random_engine arbitrarily changes algorithms. As it's intended to be used only for lightweight randomnes, it gives minimalistic garantees that the more dimensions, the more farther from standard deviation in some circumstances. To make consistency behaviour with interface compatibility, I recommend you to explicitly use std::mt19937 instead.

Reply:
https://github.com/at-wat/mcl_3dl/issues/202

Reply:
great! thanks.